File: tui/llm_terminal/session/compressor.py
from __future__ import annotations

import json

from llm_terminal.llm.prompts import (
    COMPRESSION_SUMMARY_PROMPT,
    FILESYSTEM_SNAPSHOT_PROMPT,
    SYSTEM_PROMPT,
)
from llm_terminal.session.state import ShellState


# Rough token estimate: 1 token ≈ 4 chars
def _estimate_tokens(messages: list[dict]) -> int:
    total = len(SYSTEM_PROMPT) // 4
    for msg in messages:
        content = msg.get("content", "")
        if isinstance(content, str):
            total += len(content) // 4
        elif isinstance(content, list):
            for block in content:
                total += len(str(block)) // 4
    return total


# Context window is 200k tokens; we act at 80% and 90%
_SOFT_LIMIT = 160_000   # 80% — sliding window + summary
_HARD_LIMIT = 180_000   # 90% — full snapshot reset


async def maybe_compress(
    messages: list[dict],
    state: ShellState,
    llm_complete,  # callable: (messages, system) -> str
) -> list[dict]:
    """Return a (possibly compressed) copy of the messages list."""
    tokens = _estimate_tokens(messages)

    if tokens < _SOFT_LIMIT:
        return messages

    if tokens >= _HARD_LIMIT:
        return await _hard_reset(messages, state, llm_complete)

    return await _soft_compress(messages, state, llm_complete)


async def _soft_compress(
    messages: list[dict],
    state: ShellState,
    llm_complete,
) -> list[dict]:
    """Summarize old turns, keep last 20 verbatim."""
    keep_tail = 20
    if len(messages) <= keep_tail:
        return messages

    old = messages[:-keep_tail]
    recent = messages[-keep_tail:]

    summary_text = await llm_complete(
        messages=[
            {"role": "user", "content": COMPRESSION_SUMMARY_PROMPT + "\n\n" + _flatten(old)}
        ],
        system="You are a precise summarizer. Output only valid JSON.",
    )

    try:
        summary = json.loads(summary_text.strip())
    except json.JSONDecodeError:
        # If parsing fails, just drop old messages to save context
        summary = {"note": "summary unavailable"}

    synthetic = {
        "role": "user",
        "content": f"[SESSION HISTORY SUMMARY]\n{json.dumps(summary, indent=2)}\n[END SUMMARY]",
    }

    return [synthetic] + recent


async def _hard_reset(
    messages: list[dict],
    state: ShellState,
    llm_complete,
) -> list[dict]:
    """Full filesystem snapshot, wipe history, keep last 5 turns."""
    keep_tail = 5
    recent = messages[-keep_tail:] if len(messages) > keep_tail else messages

    snapshot_text = await llm_complete(
        messages=messages + [
            {"role": "user", "content": FILESYSTEM_SNAPSHOT_PROMPT}
        ],
        system=SYSTEM_PROMPT,
    )

    try:
        snapshot = json.loads(snapshot_text.strip())
    except json.JSONDecodeError:
        snapshot = {"raw": snapshot_text[:2000]}

    synthetic = {
        "role": "user",
        "content": f"[FULL SESSION SNAPSHOT]\n{json.dumps(snapshot, indent=2)}\n[END SNAPSHOT]",
    }

    return [synthetic] + recent


def _flatten(messages: list[dict]) -> str:
    parts = []
    for msg in messages:
        role = msg.get("role", "?")
        content = msg.get("content", "")
        if isinstance(content, list):
            content = " ".join(str(b) for b in content)
        parts.append(f"{role.upper()}: {content}")
    return "\n\n".join(parts)

File: frontend/src/lib/compressor.ts
import type { Message } from '../types';
import type { ShellState } from './shellState';
import { COMPRESSION_SUMMARY_PROMPT, FILESYSTEM_SNAPSHOT_PROMPT, SYSTEM_PROMPT } from './prompts';

// Rough token estimate: 1 token ≈ 4 chars
export function estimateTokens(messages: Message[]): number {
  let total = SYSTEM_PROMPT.length / 4;
  for (const msg of messages) {
    total += msg.content.length / 4;
  }
  return Math.floor(total);
}

// Context window limits — same as Python version
const SOFT_LIMIT = 100_000; // 50% — compress early while state is still fresh
const HARD_LIMIT = 180_000; // 90% — full snapshot reset

type LLMComplete = (
  messages: Message[],
  system: string
) => Promise<{ text: string; usage?: string }>;

export async function maybeCompress(
  messages: Message[],
  state: ShellState,
  llmComplete: LLMComplete
): Promise<Message[]> {
  const tokens = estimateTokens(messages);
  if (tokens < SOFT_LIMIT) return messages;
  if (tokens >= HARD_LIMIT) return hardReset(messages, state, llmComplete);
  return softCompress(messages, state, llmComplete);
}

function flatten(messages: Message[]): string {
  return messages.map((m) => `${m.role.toUpperCase()}: ${m.content}`).join('\n\n');
}

/** Merge synthetic prepended message to avoid adjacent user messages (Gemini requirement). */
function prependSynthetic(synthetic: Message, recent: Message[]): Message[] {
  if (recent.length > 0 && recent[0].role === 'user') {
    const merged: Message = {
      role: 'user',
      content: synthetic.content + '\n\n' + recent[0].content,
    };
    return [merged, ...recent.slice(1)];
  }
  return [synthetic, ...recent];
}

async function softCompress(
  messages: Message[],
  _state: ShellState,
  llmComplete: LLMComplete
): Promise<Message[]> {
  const keepTail = 5;
  if (messages.length <= keepTail) return messages;

  const old = messages.slice(0, -keepTail);
  const recent = messages.slice(-keepTail);

  let summaryText: string;
  try {
    const result = await llmComplete(
      [{ role: 'user', content: COMPRESSION_SUMMARY_PROMPT + '\n\n' + flatten(old) }],
      'You are a precise summarizer. Output only valid JSON.'
    );
    summaryText = result.text;
  } catch {
    summaryText = '{"note":"summary unavailable"}';
  }

  let summary: unknown;
  try {
    summary = JSON.parse(summaryText.trim());
  } catch {
    summary = { note: 'summary unavailable' };
  }

  const synthetic: Message = {
    role: 'user',
    content: `[SESSION HISTORY SUMMARY]\n${JSON.stringify(summary, null, 2)}\n[END SUMMARY]`,
  };

  return prependSynthetic(synthetic, recent);
}

async function hardReset(
  messages: Message[],
  _state: ShellState,
  llmComplete: LLMComplete
): Promise<Message[]> {
  const keepTail = 3;
  const recent = messages.length > keepTail ? messages.slice(-keepTail) : [...messages];

  let snapshotText: string;
  try {
    const result = await llmComplete(
      [...messages, { role: 'user', content: FILESYSTEM_SNAPSHOT_PROMPT }],
      SYSTEM_PROMPT
    );
    snapshotText = result.text;
  } catch {
    snapshotText = '{"raw":"snapshot unavailable"}';
  }

  let snapshot: unknown;
  try {
    snapshot = JSON.parse(snapshotText.trim());
  } catch {
    snapshot = { raw: snapshotText.slice(0, 2000) };
  }

  const synthetic: Message = {
    role: 'user',
    content: `[FULL SESSION SNAPSHOT]\n${JSON.stringify(snapshot, null, 2)}\n[END SNAPSHOT]`,
  };

  return prependSynthetic(synthetic, recent);
}
